### PR TITLE
KAFKA-6398: Return value getter based on KTable materialization status

### PR DESCRIPTION
**Pull request by guozhangwang**
This is a bug fix that is composed of two parts:

1. The major part is, for all operators that is generating a KTable, we should construct its value getter based on whether the KTable itself is materialized.
1.a If yes, then query the materialized store directly for value getter.
1.b If not, then hand over to its parents value getter (recursively) and apply the computation to return.

2. The minor part is, in KStreamImpl, when joining with a table, we should connect with table's `valueGetterSupplier().storeNames()`, not the `internalStoreName()` as the latter always assume that the KTable is materialized, but that is not always true.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by guozhangwang** [comment]
@mjsax @bbejeck 

I am still working on adding the unit tests to cover these scenarios, but the fix itself should be complete. Would be appreciate if you can take a look first.

**Comment by dguy** [comment]
retest this please

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableFilter.java [hunk 1]
```diff
@@ -108,24 +92,63 @@ public void process(K key, Change<V> change) {
...
         }
 
+        @SuppressWarnings("unchecked")
         @Override
         public void init(ProcessorContext context) {
-            parentGetter.init(context);
+            if (queryableName != null) {
+                store = (KeyValueStore<K, V>) context.getStateStore(queryableName);
+            } else {
+                parentGetter.init(context);
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableFilter.java]: Could be a gap in my understanding, but wouldn't the `parentGetter` already have been initialized?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorTopology.java [hunk 2]
```diff
@@ -197,4 +197,14 @@ public String toString(final String indent) {
         return sb.toString();
     }
 
+    // for testing only
+    public Set<String> processorConnectedStateStores(final String processorName) {
+        for (ProcessorNode<?, ?> node : processorNodes) {
+            if (node.name().equals(processorName)) {
+                return node.stateStores;
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorTopology.java]: although only public for testing maybe return `Collections.unmodifiableSet(new HashSet<>(node.stateStores));` just as a precaution?

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java [hunk 3]
```diff
@@ -69,6 +73,30 @@ public void shouldAllowJoinUnmaterializedFilteredKTable() {
         builder.<Bytes, String>stream("stream-topic").join(filteredKTable, MockValueJoiner.TOSTRING_JOINER);
 
         driver.setUp(builder, TestUtils.tempDirectory());
+
+        ProcessorTopology topology = builder.internalTopologyBuilder.build();
+
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:76]: nit: final and below in other tests

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java [hunk 4]
```diff
@@ -77,14 +105,82 @@ public void shouldAllowJoinUnmaterializedMapValuedKTable() {
...
 
     @Test
     public void shouldAllowJoinMaterializedSourceKTable() {
-        final KTable<Bytes, String> table = builder.<Bytes, String>table("table-topic");
+        final KTable<Bytes, String> table = builder.table("table-topic");
         builder.<Bytes, String>stream("stream-topic").join(table, MockValueJoiner.TOSTRING_JOINER);
 
         driver.setUp(builder, TestUtils.tempDirectory());
+
+        System.out.println(builder.build().describe());
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java]: leftover debugging?

**Comment by bbejeck** [comment]
left one question and a few minor comments

**Comment by bbejeck** [comment]
failure is related - unused import in `StreamsBuilderTest`

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableFilter.java [hunk 5]
```diff
@@ -108,24 +92,63 @@ public void process(K key, Change<V> change) {
...
+            };
+        }
+    }
 
+    private class KTableFilterValueGetter implements KTableValueGetter<K, V> {
         private final KTableValueGetter<K, V> parentGetter;
+        private KeyValueStore<K, V> store = null;
 
-        public KTableFilterValueGetter(KTableValueGetter<K, V> parentGetter) {
+        KTableFilterValueGetter(KTableValueGetter<K, V> parentGetter) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableFilter.java]: nit: add `final`

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableAggregate.java [hunk 6]
```diff
@@ -33,7 +33,7 @@
 
     private boolean sendOldValues = false;
 
-    public KTableAggregate(String storeName, Initializer<T> initializer, Aggregator<? super K, ? super V, T> add, Aggregator<? super K, ? super V, T> remove) {
+    KTableAggregate(String storeName, Initializer<T> initializer, Aggregator<? super K, ? super V, T> add, Aggregator<? super K, ? super V, T> remove) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableAggregate.java]: nit: add `final` to all parameters

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableFilter.java [hunk 7]
```diff
@@ -108,24 +92,63 @@ public void process(K key, Change<V> change) {
...
             this.parentGetter = parentGetter;
         }
 
+        @SuppressWarnings("unchecked")
         @Override
         public void init(ProcessorContext context) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableFilter.java]: nit: add `final`

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java [hunk 8]
```diff
@@ -21,29 +21,66 @@
...
+        return new KTableKTableJoinMergeProcessor();
     }
 
     @Override
     public KTableValueGetterSupplier<K, V> view() {
-        return parent1.valueGetterSupplier();
+        // if the result KTable is materialized, use the materialized store to return getter value;
+        // otherwise rely on the parent getter and apply join on-the-fly
+        if (queryableName != null) {
+            return new KTableValueGetterSupplier<K, V>() {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java]: This is the same code as in `KTableFilter` -- we should refactor and share code.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java [hunk 9]
```diff
@@ -21,29 +21,66 @@
...
+                public KTableValueGetter<K, V> get() {
+                    return parent1.valueGetterSupplier().get();
+                }
+
+                @Override
+                public String[] storeNames() {
+                    // we need to allow the downstream processor to be able to access both ends of the joining table's value getters
+                    final String[] storeNames1 = parent1.valueGetterSupplier().storeNames();
+                    final String[] storeNames2 = parent2.valueGetterSupplier().storeNames();
+                    final ArrayList<String> stores = new ArrayList<>(storeNames1.length + storeNames2.length);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java]: Why don't using an array from the beginning on?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java [hunk 10]
```diff
@@ -53,8 +90,7 @@ public void enableSendingOldValues() {
         sendOldValues = true;
     }
 
-    private class KTableKTableJoinMergeProcessor<K, V>
-        extends AbstractProcessor<K, Change<V>> {
+    private class KTableKTableJoinMergeProcessor extends AbstractProcessor<K, Change<V>> {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java:86]: Why do you remove generics types?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java [hunk 11]
```diff
@@ -21,29 +21,66 @@
...
+                           final String queryableName) {
         this.parent1 = parent1;
         this.parent2 = parent2;
         this.queryableName = queryableName;
     }
 
     @Override
     public Processor<K, Change<V>> get() {
-        return new KTableKTableJoinMergeProcessor<>();
+        return new KTableKTableJoinMergeProcessor();
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java:41]: Why this change? Compare my comment below.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableMapValues.java [hunk 12]
```diff
@@ -44,19 +45,34 @@ public KTableMapValues(final KTableImpl<K, ?, V> parent, final ValueMapperWithKe
...
-
-            @Override
-            public String[] storeNames() {
-                return parentValueGetterSupplier.storeNames();
-            }
-        };
+        // if the KTable is materialized, use the materialized store to return getter value;
+        // otherwise rely on the parent getter and apply map-values on-the-fly
+        if (queryableName != null) {
+            return new KTableValueGetterSupplier<K, V1>() {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableMapValues.java]: Can be shared bode with `Filter` and `Merger`

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableMapValues.java [hunk 13]
```diff
@@ -44,19 +45,34 @@ public KTableMapValues(final KTableImpl<K, ?, V> parent, final ValueMapperWithKe
...
+                    return new KTableMaterializedValueGetter<>(queryableName);
+                }
+
+                @Override
+                public String[] storeNames() {
+                    return new String[]{queryableName};
+                }
+            };
+        } else {
+            return new KTableValueGetterSupplier<K, V1>() {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableMapValues.java]: This is the same as for `Filter` -- we should refactor and share code.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableMaterializedValueGetter.java [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+    private final String storeName;
+    private KeyValueStore<K, V> store;
+
+    KTableMaterializedValueGetter(final String storeName) {
+        this.storeName = storeName;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public void init(ProcessorContext context) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableMaterializedValueGetter.java]: nit: add `final`

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableMaterializedValueGetter.java [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+        store = (KeyValueStore<K, V>) context.getStateStore(storeName);
+    }
+
+    @Override
+    public V get(K key) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableMaterializedValueGetter.java]: nit: add `final`

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableReduce.java [hunk 16]
```diff
@@ -31,7 +31,7 @@
 
     private boolean sendOldValues = false;
 
-    public KTableReduce(String storeName, Reducer<V> addReducer, Reducer<V> removeReducer) {
+    KTableReduce(String storeName, Reducer<V> addReducer, Reducer<V> removeReducer) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableReduce.java]: nit: add `final` to all parameters

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableRepartitionMap.java [hunk 17]
```diff
@@ -33,7 +33,7 @@
     private final KTableImpl<K, ?, V> parent;
     private final KeyValueMapper<? super K, ? super V, KeyValue<K1, V1>> mapper;
 
-    public KTableRepartitionMap(KTableImpl<K, ?, V> parent, KeyValueMapper<? super K, ? super V, KeyValue<K1, V1>> mapper) {
+    KTableRepartitionMap(KTableImpl<K, ?, V> parent, KeyValueMapper<? super K, ? super V, KeyValue<K1, V1>> mapper) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableRepartitionMap.java]: nit: add `final` to all parameters

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableRepartitionMap.java [hunk 18]
```diff
@@ -101,7 +101,7 @@ public void process(K key, Change<V> change) {
 
         private final KTableValueGetter<K, V> parentGetter;
 
-        public KTableMapValueGetter(KTableValueGetter<K, V> parentGetter) {
+        KTableMapValueGetter(KTableValueGetter<K, V> parentGetter) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableRepartitionMap.java]: nit: add `final`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorTopology.java [hunk 19]
```diff
@@ -197,4 +197,14 @@ public String toString(final String indent) {
         return sb.toString();
     }
 
+    // for testing only
+    public Set<String> processorConnectedStateStores(final String processorName) {
+        for (ProcessorNode<?, ?> node : processorNodes) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorTopology.java]: nit: add `final`

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java [hunk 20]
```diff
@@ -69,6 +73,30 @@ public void shouldAllowJoinUnmaterializedFilteredKTable() {
         builder.<Bytes, String>stream("stream-topic").join(filteredKTable, MockValueJoiner.TOSTRING_JOINER);
 
         driver.setUp(builder, TestUtils.tempDirectory());
+
+        ProcessorTopology topology = builder.internalTopologyBuilder.build();
+
+        assertThat(topology.stateStores().size(), equalTo(1));
+        assertThat(topology.processorConnectedStateStores("KSTREAM-JOIN-0000000005").size(), is(1));
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java]: This an the one below can be merged into a single assertThat with "equals" and a singelton Set as expected value

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java [hunk 21]
```diff
@@ -69,6 +73,30 @@ public void shouldAllowJoinUnmaterializedFilteredKTable() {
...
+        assertThat(topology.processorConnectedStateStores("KSTREAM-JOIN-0000000005"), hasItem(topology.stateStores().get(0).name()));
+        assertThat(topology.processorConnectedStateStores("KTABLE-FILTER-0000000003").size(), is(0));
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java]: use `isEmpty` instead of size=0 ?

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java [hunk 22]
```diff
@@ -69,6 +73,30 @@ public void shouldAllowJoinUnmaterializedFilteredKTable() {
...
+
+        driver.setUp(builder, TestUtils.tempDirectory());
+
+        ProcessorTopology topology = builder.internalTopologyBuilder.build();
+
+        assertThat(topology.stateStores().size(), equalTo(2));
+        assertThat(topology.processorConnectedStateStores("KSTREAM-JOIN-0000000005").size(), is(1));
+        assertThat(topology.processorConnectedStateStores("KSTREAM-JOIN-0000000005"), hasItem("store"));
+        assertThat(topology.processorConnectedStateStores("KTABLE-FILTER-0000000003").size(), is(1));
+        assertThat(topology.processorConnectedStateStores("KTABLE-FILTER-0000000003"), hasItem("store"));
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java]: as above: use equals plus singleton set both both

**Review by mjsax — COMMENTED** [review]
Thanks for the PR.

You mention that we would go back recursively if we have non-materialized tables, but we don't have a test for this case yet.

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java (hunk 4 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java]: Ack.

**Comment by guozhangwang** [comment]
> You mention that we would go back recursively if we have non-materialized tables, but we don't have a test for this case yet.

The newly added test in `StreamsBuilderTest` covers this: the follow-up join processor should be able to access the up-stream processor's associated state stores.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java (hunk 10 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java:86]: It is actually not needed: note that this is a private class that is only used for returning the processor, and hence can completely inherit the generics from its parent `KTableKTableJoinMerger`.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java (hunk 11 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java:41]: See my comment below.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java (hunk 9 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java]: This is to be coherent with `Topology#connectProcessorAndStateStores` with array store names.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java (hunk 8 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java]: Ack. I'm going to enlarge `KTableMaterializedValueGetter` to `KTableMaterializedValueGetterSupplier`

**Comment by guozhangwang** [comment]
@mjsax @bbejeck comments addressed.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java (hunk 9 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java]: Cannot follow. We need to return an array be create a ArryaList that we convert in an array later. Why not an array directly instead of an ArrayList?

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java]: `storeNames` is used in `connectProcessorAndStateStores`, which can also be called by user to connect one processor with multiple state stores, where state store names can be passed as a enum array.

**Comment by guozhangwang** [comment]
retest this please

**Comment by guozhangwang** [comment]
retest this please

**Comment by guozhangwang** [comment]
Re-opened for the jenkins failure, which seems not related.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java (hunk 9 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java]: I understand that. I am just wondering, why we create a `ArrayList` here in stead of a plain array:
```
final String[] stores = new String[storeNames1.length + storeNames2.length];
int i = 0;
for (int j = 0; j < storeNames1.length; ++j, ++i) {
    stores[i] = storeNames1[j]
}
for (int j = 0; j < storeNames2.length; ++j, ++i) {
    stores[i] = storeNames2[j]
}
return stores;
```

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java]: Ah I see what you mean, ack.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableFilter.java [hunk 23]
```diff
@@ -108,24 +92,45 @@ public void process(K key, Change<V> change) {
...
 
-    private class KTableFilterValueGetter implements KTableValueGetter<K, V> {
+    @Override
+    public KTableValueGetterSupplier<K, V> view() {
+        // if the KTable is materialized, use the materialized store to return getter value;
+        // otherwise rely on the parent getter and apply filter on-the-fly
+        if (queryableName != null) {
+            return new KTableMaterializedValueGetterSupplier<>(queryableName);
+        } else {
+            return new KTableValueGetterSupplier<K, V>() {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableFilter.java:102]: I would really like to share code with `KTableMapValues` . Can't we create an interface for the `computeValue()` method that both `KTableFilterValueGetter` and `KTableMapValueValueGetter` need to call? It's the exact some pattern for MapValue and Filter.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableMaterializedValueGetterSupplier.java [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+        public void init(final ProcessorContext context) {
+            store = (KeyValueStore<K, V>) context.getStateStore(storeName);
+        }
+
+        @Override
+        public V get(final K key) {
+            return store.get(key);
+        }
+    }
+
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableMaterializedValueGetterSupplier.java]: nit: remove one empty line

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableFilter.java (hunk 23 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableFilter.java:102]: I tried to do that, but it ended up not saving much LOC since we need to have a new interface class with both of them implementing it. So I ended up leaving this as is.

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableFilter.java:102]: Ok. Fine to leave as-is.

**Comment by mjsax** [comment]
Checkstyle error:  unused imports in `KTableKTableJoinMerger`

**Comment by guozhangwang** [comment]
> Checkstyle error: unused imports in KTableKTableJoinMerger

Hmm. should be fixed in my last commit?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java [hunk 25]
```diff
@@ -28,22 +28,52 @@
...
     }
 
     @Override
     public KTableValueGetterSupplier<K, V> view() {
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java:45]: Would it be possible to add a test for this class or do you feel it is covered elsewhere?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java [hunk 26]
```diff
@@ -28,22 +28,52 @@
...
-        return parent1.valueGetterSupplier();
+        // if the result KTable is materialized, use the materialized store to return getter value;
+        // otherwise rely on the parent getter and apply join on-the-fly
+        if (queryableName != null) {
+            return new KTableMaterializedValueGetterSupplier<K, V>(queryableName);
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java]: nit: don't need `<K, V>` can just be `<>`

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java [hunk 27]
```diff
@@ -28,22 +28,52 @@
...
+            return new KTableValueGetterSupplier<K, V>() {
+
+                public KTableValueGetter<K, V> get() {
+                    return parent1.valueGetterSupplier().get();
+                }
+
+                @Override
+                public String[] storeNames() {
+                    // we need to allow the downstream processor to be able to access both ends of the joining table's value getters
+                    final String[] storeNames1 = parent1.valueGetterSupplier().storeNames();
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java:60]: Why not something like:
```
final List<String> storeNames = Arrays.asList(parent1.valueGetterSupplier().storeNames());
                    storeNames.addAll(Arrays.asList(parent2.valueGetterSupplier().storeNames()));
                    return storeNames.toArray(new String[storeNames.size()]);
```
?
I don't think it is on the critical path so performance shouldn't be an issue

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableMapValues.java [hunk 28]
```diff
@@ -44,19 +45,24 @@ public KTableMapValues(final KTableImpl<K, ?, V> parent, final ValueMapperWithKe
...
-            public KTableValueGetter<K, V1> get() {
-                return new KTableMapValuesValueGetter(parentValueGetterSupplier.get());
-            }
-
-            @Override
-            public String[] storeNames() {
-                return parentValueGetterSupplier.storeNames();
-            }
-        };
+        // if the KTable is materialized, use the materialized store to return getter value;
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableMapValues.java:48]: This looks indentical to the same in `KTableFilter` - can we not have another class that they both call that does this?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorTopology.java [hunk 29]
```diff
@@ -197,4 +197,14 @@ public String toString(final String indent) {
         return sb.toString();
     }
 
+    // for testing only
+    public Set<String> processorConnectedStateStores(final String processorName) {
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorTopology.java:201]: Hmm - is there another way we can test this? Don't really like having methods that are purely for testing

**Review by dguy — COMMENTED** [review]
Thanks @guozhangwang, left a few comments

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java (hunk 26 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java]: Ack.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java (hunk 27 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java:60]: What you suggested is actually the original code, while @mjsax suggested doing this. I am actually fine either way.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableMapValues.java (hunk 28 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableMapValues.java:48]: @mjsax has the same comment, and I replied here: https://github.com/apache/kafka/pull/4421#discussion_r162220745

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorTopology.java (hunk 29 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorTopology.java:201]: I have thought about this, via `TopologyDescription` but the code path would be very awkward. I ended up adding this function mainly because `ProcessorTopology` is an internal class so I felt it could be fine.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java (hunk 25 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java:45]: I think it is covered in `KTableKTableJoinTest#testJoin`. And I have created another JIRA for its follow-up, to let all such unit test to use a mock state store to record num. get / put calls: https://issues.apache.org/jira/browse/KAFKA-6460

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java (hunk 27 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java:60]: Damian suggest is actually different as it avoid the loops... That would also be fine. But using ArrayList plus for-loops is a mess IMHO.

**Review by mjsax — APPROVED** [review]
LGTM.

**Comment by guozhangwang** [comment]
Merged to trunk and 1.0.
